### PR TITLE
Draft: new fresh helm chart

### PR DESCRIPTION
**Pull request by robjuz**
Hallo @tobybatch 
First of all thanks for your work.
Unfortunately I was unable do deploy kimai without issues. That's why I created my own chart inspired by the bitnami/wordpress chart.

This chart has some improvements connected to volume mounts and Ingress.

I also included the bitnami/mariadb chart as a dependency with some reasonable default configuration. This gives the user more controller about the deployment.

I still have some work to do:

1. Address the ReadWriteOnce PVC problem when scaling up
2. Update the Readme with new values
3. Add value schema

### CODE REVIEW

**Comment by tobybatch** [comment]
@robjuz Thanks for the contrib. I, unfortunately, know diddly squat about helm.

@stranljip, can you take a quick look please and just sanity check Robjuz's PR and leave us a comment please?

File: docs/helm/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,33 @@
...
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "apache"
```
Comment (stranljip) [inline docs/helm/Chart.yaml:6]: `apache` as version number does not really make sense imho

File: docs/helm/Chart.yaml [hunk 2]
```diff
@@ -1,6 +1,33 @@
...
+# Library charts provide useful utilities or functions for the chart developer. They're included as
+# a dependency of application charts to inject those utilities and functions into the rendering
+# pipeline. Library charts do not define any templates and therefore cannot be deployed.
 type: application
-version: "0.1.0"
-appVersion: "0.1.0"
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
```
Comment (stranljip) [inline docs/helm/Chart.yaml]: I would suggest removing the default texts (also below)

File: docs/helm/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
+      containers:
+        - name: {{ .Chart.Name }}
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
```
Comment (stranljip) [inline docs/helm/templates/deployment.yaml:36]: OK, now I understand why you used `apache` as version.

File: docs/helm/Chart.yaml (hunk 1 above)
Comment (stranljip) [reply docs/helm/Chart.yaml:6]: OK - after seeing how you use this for getting the image tag, this makes sense.

File: docs/helm/values.yaml [hunk 4]
```diff
@@ -1,91 +1,164 @@
-# Values for the Helm chart for Kimai
+# Default values for kimai2.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
```
Comment (stranljip) [inline docs/helm/values.yaml]: I suggest removing the default text/comments in order to make the values better readable (also further on in this file)

File: docs/helm/values.yaml [hunk 5]
```diff
@@ -1,91 +1,164 @@
...
-  kimai:
-    # -- Host name for Kimai. Defaults to kimai
-    hostName:
-    # -- Optional possibility to declare a specific domain for Kimai
-    domain:
-    # -- Certificate for Kimai
-    tlsSecret: add-you-own-here
+  enabled: false
+  annotations:
+    kubernetes.io/ingress.class: nginx
```
Comment (stranljip) [inline docs/helm/values.yaml]: this annotation is deprecated. If you need it for your use case, I would suggest to allow extra labels/annotations to be specified optionally.

File: docs/helm/values.yaml [hunk 6]
```diff
@@ -1,91 +1,164 @@
...
+    kubernetes.io/tls-acme: "true"
```
Comment (stranljip) [inline docs/helm/values.yaml]: AFAIK this is a very specific way to request a certificate. I would suggest to leave out something like this and handle the TLS certificate generation outside of the chart

**Review by stranljip — COMMENTED** [review]
Hi @robjuz - great work, the chart looks definitely more consolidated now! Personally I prefer to handle all dependencies in my own chart instead of relying on dependencies, but this is probably fine. One thing I saw in general is that there are some "default texts" in a number of places which I would rather delete in order to keep the files concise. 
I have not tested deploying the chart, only went through the files and added my thoughts.

**Comment by wagnst** [comment]
@robjuz will you review so it can be merged? Waiting for some "official" k8s deployments.

**Comment by robjuz** [comment]
@wagnst I will, soon.

**Comment by robjuz** [comment]
@stranljip I made some changes according to your remarks. I also updated the readme. I was heavily inspired by the bitnami/wordpress chart. I thing it looks ok now.

You could also consider using github-pages as a helm chart repo. I'm doing it with my own (2 for now ;) chart collection.

https://github.com/robjuz/helm-charts/blob/master/.github/workflows/release.yaml

**Comment by tobybatch** [comment]
@robjuz @stranljip I think the suggestion of setting up a repo to handle the helm charts for this project is a great idea. I'm loathed to merge in stuff I have no idea what it does, particularly is there is some _opinion_ in what is the right way of doing it.

I'd link to it in the docs and then you could manage your own PRs. 

Ideas?

**Comment by kevinpapst** [comment]
The Kimai docs are open for PRs as well: https://github.com/kimai/www.kimai.org/tree/master/_documentation

**Comment by tobybatch** [comment]
Oh yeah, that would work too. @kevinpapst my problem with the helm stuff is a bit like yours with docker. I want to support it but I just don't use it so I'm a bit _at sea_ when it come to reviewing.

**Comment by robjuz** [comment]
As I mentioned, I deployed kimai with this chart.
You can also take a look at https://github.com/robjuz/helm-charts/tree/master/charts/kimai2
Maybe @wagnst could try it out and give feedback. I'm also not a helm/kubernetes expert.

**Comment by neerdoc** [comment]
@robjuz Hey, nice work with the helm repo. For some reason I can't write an issue at [helm-charts](https://github.com/robjuz/helm-charts/tree/master/charts/kimai2), but I also tried your chart from my rancher installation and unfortunately it fails "Error: found in Chart.yaml, but missing in charts/ directory: common, mariadb"

Not an helm expert either, but I can try to help out with testing if needed!

**Comment by robjuz** [comment]
@neerdoc How do you USD the chart? You need either add the repo 

helm repo add robjuz https://robjuz.github.io/helm-charts/
helm install kimai robjuz/kimai2

or build the dependencies

https://helm.sh/docs/helm/helm_dependency_build/

**Comment by tobybatch** [comment]
@robjuz OK, I propose that I replace the helm references in the docs to a link out to https://github.com/robjuz/helm-charts/tree/master/charts/kimai2

@stranljip If you have another repo I can link to that too but it may be better if you and @robjuz work together but I don't want to rail road anyone.

This is exactly the same problem kevin and I had about dockers. Kev was really willing to support dockers but didn't use them so couldn't really manage PRs etc. So We created this repo and he just links to it.

**Comment by stranljip** [comment]
> @stranljip If you have another repo I can link to that too but it may be better if you and @robjuz work together but I don't want to rail road anyone.

No, I am fine with that.

**Comment by neerdoc** [comment]
> @neerdoc How do you USD the chart? You need either add the repo
> 
> helm repo add robjuz https://robjuz.github.io/helm-charts/
> helm install kimai robjuz/kimai2
> 
> or build the dependencies
> 
> https://helm.sh/docs/helm/helm_dependency_build/

Ah, I had linked directly to the github repo! Changing to the helm generated chart worked better!

Now it created PVCs and the database starts! However, I have new issues... But is there a slack channel or whatever where we can talk instead of in this thread? 

Current issue: "In Kernel.php line 688: Unable to create the "cache" directory (/opt/kimai/var/cache/prod)."

**Comment by robjuz** [comment]
I enabled issues in my repo. Please create one :)

W dniu śr., 26.05.2021 o 10:47 Gustav Johansson ***@***.***>
napisał(a):

**Comment by tobybatch** [comment]
Excellent, thanks everyone, I'll close this and raise a ticket for me to update the docs to point to the kimai page of @robjuz 's repo

**Pull request closed without merging**